src/__test__/components/plots-and-tables/markerHeatmap.test.jsx

```diff
@@ -0,0 +1,235 @@
+      status: {
+        pipeline: {
+          startDate: '2020-01-01T00:00:00',
+          status: 'SUCEEDED',
+        },
+        worker: { status: 'Running' },
+        gem2s: { status: 'SUCCEEDED' },
+      },
+    },
+    processing: {
+      configureEmbedding: {
+        clusteringSettings: {
+          methodSettings: {
+            louvain: {
+              resolution: 0.8,
+            },
+          },
+        },
+      },
+    },
+  },
+  cellSets: {
+    loading: false,
+    hierarchy: [
+      {
+        key: 'louvain',
+        children: [
+          {
+            key: 'louvain-0',
+          },
+          {
+            key: 'louvain-1',
+          },
+          {
+            key: 'louvain-2',
+          },
+          {
+            key: 'louvain-3',
+          },
+        ],
+      },
+    ],
+    properties: {
+      louvain: {
+        type: 'cellSets',
+        cellIds: new Set([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]),
+      },
+      'louvain-0': {
+        name: 'Cluster 0',
+        color: '#e377c2',
+        cellIds: new Set([1, 2, 3, 4]),
+        rootNode: false,
+        type: 'cellSets',
+      },
+      'louvain-1': {
+        name: 'Cluster 1',
+        color: '#8c564b',
+        cellIds: new Set([5, 6, 7, 8]),
+        rootNode: false,
+        type: 'cellSets',
+      },
+      'louvain-2': {
+        name: 'Cluster 2',
+        color: '#d62728',
+        cellIds: new Set([9, 10, 11, 12]),
+        rootNode: false,
+        type: 'cellSets',
+      },
+      'louvain-3': {
+        name: 'Cluster 3',
+        color: '#2ca02c',
+        cellIds: new Set([13, 14, 15, 16]),
+        rootNode: false,
+        type: 'cellSets',
+      },
+      views: [],
+    },
+    hidden: 'Set()',
+  },
+  genes: {
+    markers: {
+      error: false,
+    },
+    expression: {
+      loading: [],
+
+      data: {
+        gene0: {
+          rawExpression: { expression: [5, 5, 5, 5, 0, 4, 4, 4, 2, 3, 4, 5, 0, 0, 4, 5] },
+        },
+        gene1: {
+          rawExpression: { expression: [1, 2, 3, 4, 5, 4, 5, 5, 3, 0, 0, 0, 0, 0, 0, 0] },
+        },
+        gene3: {
+          rawExpression: { expression: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 0, 1, 0] },
+        },
+      },
+      views: {
+        [plotUuid]: {
+          data: ['gene0', 'gene1', 'gene3'],
+        },
+      },
+    },
+    properties: { views: { [plotUuid]: [] } },
+  },
+};
+
+const experimentId = 'randomExperiment';
+let store = null;
+let loadConfigSpy = null;
+let loadMarkersSpy;
+let configUpdatedSpy;
+let fetchCachedWorkSpy;
+
+const renderHeatmapPage = async () => {
+  store = createStore(rootReducer, _.cloneDeep(defaultStore), applyMiddleware(thunk));
+
+  rtl.render(
+    <Provider store={store}>
+      <MarkerHeatmap
+        experimentId={experimentId}
+      />
+    </Provider>,
+  );
+  await rtl.waitFor(() => expect(loadConfigSpy).toHaveBeenCalledTimes(1));
+};
+
+describe('Marker heatmap plot', () => {
```

Comment:
that's some good addition of tests, but it is not sufficient. Add tests to check for what happens if the redux store is empty, if parts of the redux store are empty, if after requesting expression an error is returned and so on. Try to think what can possibly go wrong in this file and add a test for it.

Reply:
There are also lots of small functions that are not covered like `getCellClusterFromCellId` -- we need them tested too. Looking at the function `getCellClusterFromCellId`, looks like it will crash if `properties` is undefined. We have to have guards against all these things and make sure that the code works not only if the user uses it exactly as they are intended to, but when they try to do something else. Try to apply a mindset towards programming for failure -- go through each line and think: "how can possibly this line break?" and then add handling for that and a unit test.

Reply:
These functions are tested - they are used when sorting, which is tested.